src/clarity-angular/utils/virtual-scroll/virtual-for-of.ts

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2016-2017 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+import {
+    AfterViewInit,
+    ChangeDetectorRef,
+    Directive,
+    EmbeddedViewRef,
+    Input,
+    Renderer2,
+    TemplateRef,
+    ViewContainerRef
+} from "@angular/core";
+
+import {
+    appendAndPreserveScroll,
+    offsetHeight,
+    prependAndPreserveScroll,
+    ratioBottomReady,
+    ratioTopReady
+} from "./dom-helpers";
+import {isNonNgIterable, NonNgIterable} from "./non-ng-iterable";
+
+enum Side {
+    START = -1,
+    END = 1
+}
+
+@Directive({selector: "[clrVirtualForOf]"})
+export class VirtualForOf<T> implements AfterViewInit {
+    constructor(private viewContainer: ViewContainerRef, private template: TemplateRef<{$implicit: T}>,
+                private renderer: Renderer2, private cdr: ChangeDetectorRef) {
+        this.containerEl = <HTMLElement>renderer.parentNode(viewContainer.element.nativeElement);
+    }
+
+    private containerEl: HTMLElement;
+
+    // Until we have a way to tell Angular not to iterate over the whole iterable as soon as it's passed as an input,
+    // we can't accept an NgIterable here so we have this weird Array | NonNgIterable type.
+    private _items: Array<T>|NonNgIterable<T>;
+    @Input("clrVirtualForOf")
+    set items(value: Array<T>|NonNgIterable<T>) {
+        this._items = value;
+        this.viewContainer.clear();
+        this.fill();
```

Comment:
You asked me to remind you to add:

```
this.start = -1;
this.end = 0;
```

before

`this.fill()`.


Blocking PR to make sure that you read my comment :P